prodigy_hf/textcat.py

```diff
@@ -0,0 +1,236 @@
+import random
+import time
+from pathlib import Path
+from typing import Dict, Iterable, List, Literal, Optional
+
+import evaluate
+import numpy as np
+from datasets import Dataset, DatasetDict
+from datasets.utils.logging import disable_progress_bar
+from prodigy.components.db import connect
+from prodigy.components.stream import get_stream
+from prodigy.core import Arg, recipe
+from prodigy.util import log
+from transformers import (
+    AutoModelForSequenceClassification,
+    AutoTokenizer,
+    DataCollatorWithPadding,
+    Trainer,
+    TrainingArguments,
+    pipeline
+)
+from transformers.utils.logging import set_verbosity_error as set_transformers_verbosity_error
+
+
+def get_label_names(examples: List[Dict], variant: Literal["binary", "multi"]) -> List[str]:
+    """We have to assume exclusive textcat here. So the first example contains all labels."""
+    if variant == "multi":
+        return [ex['id'] for ex in examples[0]['options']]
+    return ["accept", "reject"]
+
+
+def into_hf_format(train_examples: List[Dict], valid_examples: List[Dict], variant: Literal["binary", "multi"]):
+    """Turn the examples into variables/format that Huggingface expects."""
+    label_names = get_label_names(train_examples, variant)
+    id2label = {i: n for i, n in enumerate(label_names)}
+    label2id = {n: i for i, n in enumerate(label_names)}
+
+    def generator(examples) -> Iterable[Dict]:
+        for ex in examples:
+            label = None
+            if variant == "binary":
```

Comment:
To be able to use it in the `correct` recipe, maybe it would make more sense to use the following label assignment logic: if accept, use the ex["label"], elif reject use NOT+ex["label"], else skip. Accept and Reject as labels won't make much sense in `correct`, right?

Reply:
As discussed during our huddle: we'll not allow for binary HF models in `*.correct` recipes for now. Just to make sure we don't resort to flaky assumptions because this path will require us to make decisions based on string matching.